pkg/hydrator/nexpose.go

```diff
@@ -2,38 +2,324 @@ package hydrator
+	if err != nil {
+		return "", err
+	}
+	return solutionDetails.Steps.Text, nil
+}
+
+// FetchVulnerabilityDetails fetches details about a particular vulnerability
+func (n *NexposeClient) FetchVulnerabilityDetails(ctx context.Context, vulnID string) (NexposeVulnerability, error) {
+	body, err := n.makeNexposeRequest(nil, "api", "3", "vulnerabilities", vulnID)
+	if err != nil {
+		return NexposeVulnerability{}, err
+	}
+	var vulnDetails vulnerabilityDetails
+	err = json.Unmarshal(body, &vulnDetails)
+	if err != nil {
+		return NexposeVulnerability{}, err
+	}
+	return NexposeVulnerability{
+		Title:          vulnDetails.Title,
+		Description:    vulnDetails.Description.Text,
+		CvssV2Score:    vulnDetails.CVSS.V2.Score,
+		CvssV2Severity: cvssV2Severity(vulnDetails.CVSS.V2.Score),
+	}, nil
+}
+
+// FetchAssetVulnerabilities fetches the list of vulnerabilities for an asset
+func (n *NexposeClient) FetchAssetVulnerabilities(ctx context.Context, assetID int64) ([]NexposeAssetVulnerability, error) {
+	body, err := n.makeNexposeRequest(
+		map[string]string{pageQueryParam: "0", sizeQueryParam: strconv.Itoa(n.PageSize)},
+		"api", "3", "assets", strconv.FormatInt(assetID, 10), "vulnerabilities",
+	)
+	if err != nil {
+		return nil, err
+	}
+	var assetVulns assetVulnerabilities
+	err = json.Unmarshal(body, &assetVulns)
+	if err != nil {
+		return nil, err
+	}
+
+	pages := int(assetVulns.Page.TotalPages)
+	totalResources := int(assetVulns.Page.TotalResources)
+	assetVulnsChannel := make(chan NexposeAssetVulnerability, totalResources)
+	errorChan := make(chan error, pages)
+	for _, resource := range assetVulns.Resources {
+		assetVulnsChannel <- assetVulnToNexposeAssetVuln(resource)
+	}
+
+	for curPage := 1; curPage < pages; curPage = curPage + 1 {
+		go n.makePagedAssetVulnerabilitiesRequest(assetID, curPage, assetVulnsChannel, errorChan)
+	}
+
+	nexposeAssetVulns := make([]NexposeAssetVulnerability, 0, totalResources)
+	for i := 0; i < totalResources; i = i + 1 {
+		select {
+		case assetVuln := <-assetVulnsChannel:
+			nexposeAssetVulns = append(nexposeAssetVulns, assetVuln)
+		case err := <-errorChan:
+			return nil, err
+		}
+	}
+	return nexposeAssetVulns, nil
+}
+
+func (n *NexposeClient) makePagedAssetVulnerabilitiesRequest(assetID int64, page int, assetVulnsChannel chan NexposeAssetVulnerability, errorChan chan error) {
+	body, err := n.makeNexposeRequest(
+		map[string]string{pageQueryParam: strconv.Itoa(page), sizeQueryParam: strconv.Itoa(n.PageSize)},
+		"api", "3", "assets", strconv.FormatInt(assetID, 10), "vulnerabilities",
+	)
+	if err != nil {
+		errorChan <- err
+		return
+	}
+	var assetVulns assetVulnerabilities
+	err = json.Unmarshal(body, &assetVulns)
+	if err != nil {
+		errorChan <- err
+		return
+	}
+	for _, resource := range assetVulns.Resources {
+		assetVulnsChannel <- assetVulnToNexposeAssetVuln(resource)
+	}
+}
+
+func (n *NexposeClient) makeNexposeRequest(queryParams map[string]string, pathFragments ...string) ([]byte, error) {
```

Comment:
info: Something to consider that we likely don't need to action right now is: "How would you constrain the number of concurrent calls to this function?". This isn't something we typically spend a lot of time thinking about because our projects usually support a much higher concurrency and throughput than production traffic. However, this is a special case of making requests to a 3rd party system that may not be build to support the traffic we end up throwing at it.

When we talked about the layering pattern for adding concurrent fan-out we also talked through how we might limit concurrent invocations of any given function by wrapping it with a semaphore. If we use the same semaphore to wrap all of our interfaces that represent Nexpose calls then we would effectively limit the total number of concurrent calls for all APIs. However, at least one of the operations you have here is a fan-out operation to fetch the paginated vulnerability details. This means that one of the operations has a disproportionately higher cost than the other by the nature of it making N calls per invocation rather than one.

The thought experiment you might consider is what you would need to do if we decided, later, that we need to limit the total outgoing calls to Nexpose to a maximum of N concurrently.

Reply:
Good callout. A lot of this refactoring was an attempt to reduce the burden of making a request to nexpose (auth in particular).

I'm guessing we could create an interface for the client for making the HTTP requests to Nexpose and wrap the client with rate-limiting logic. Being that the client itself implements the interface - I would think it might be possible to implement rate limiting at the `Transport` layer in order to limit the total number of concurrent requests. The interface could help inform the granularity to which you'd have the ability to control these things.